From: Martina Ferrari <tina@debian.org>
Date: Sat, 20 Jun 2020 15:32:33 -0300
Subject: Do not embed blobs

Forwarded: not-needed
Last-Update: Mon, 14 Mar 2022 23:20:08 +0000

Avoid embedding blobs into the prometheus binary, instead use files
installed on disk.
---
 cmd/prometheus/main.go |  3 +++
 web/ui/ui.go           |  3 ++-
 web/web.go             | 51 +++++++++++++++++++++-----------------------------
 3 files changed, 26 insertions(+), 31 deletions(-)

diff --git a/cmd/prometheus/main.go b/cmd/prometheus/main.go
index 1b82cc7..41de51d 100644
--- a/cmd/prometheus/main.go
+++ b/cmd/prometheus/main.go
@@ -392,6 +392,9 @@ func main() {
 		"Prefix for the internal routes of web endpoints. Defaults to path of --web.external-url.").
 		PlaceHolder("<path>").StringVar(&cfg.web.RoutePrefix)
 
+	a.Flag("web.local-assets", "Path to static asset/templates directory.").
+		Default("/usr/share/prometheus/web/").StringVar(&cfg.web.LocalAssets)
+
 	a.Flag("web.user-assets", "Path to static asset directory, available at /user.").
 		PlaceHolder("<path>").StringVar(&cfg.web.UserAssetsPath)
 
diff --git a/web/ui/ui.go b/web/ui/ui.go
index 2585951..842dd4a 100644
--- a/web/ui/ui.go
+++ b/web/ui/ui.go
@@ -11,7 +11,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//go:build !builtinassets
+//go:build ignore
+// +build ignore
 
 package ui
 
diff --git a/web/web.go b/web/web.go
index 601d42c..6178340 100644
--- a/web/web.go
+++ b/web/web.go
@@ -17,8 +17,10 @@ import (
 	"bytes"
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io"
+	"io/fs"
 	"log/slog"
 	"math"
 	"net"
@@ -44,7 +46,6 @@ import (
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/common/promslog"
 	"github.com/prometheus/common/route"
-	"github.com/prometheus/common/server"
 	toolkit_web "github.com/prometheus/exporter-toolkit/web"
 	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
 	"go.uber.org/atomic"
@@ -60,7 +61,6 @@ import (
 	"github.com/prometheus/prometheus/util/netconnlimit"
 	"github.com/prometheus/prometheus/util/notifications"
 	api_v1 "github.com/prometheus/prometheus/web/api/v1"
-	"github.com/prometheus/prometheus/web/ui"
 )
 
 // Paths handled by the React router that should all serve the main React app's index.html,
@@ -276,7 +276,7 @@ type Options struct {
 	MaxConnections             int
 	ExternalURL                *url.URL
 	RoutePrefix                string
-	UseLocalAssets             bool
+	LocalAssets                string
 	UserAssetsPath             string
 	ConsoleTemplatesPath       string
 	ConsoleLibrariesPath       string
@@ -422,17 +422,13 @@ func New(logger *slog.Logger, o *Options) *Handler {
 		})
 	}
 
-	reactAssetsRoot := "/static/mantine-ui"
-	if h.options.UseOldUI {
-		reactAssetsRoot = "/static/react-app"
-	}
+	// reactAssetsRoot := "/static/mantine-ui"
+	// if h.options.UseOldUI {
+	// 	reactAssetsRoot = "/static/react-app"
+	// }
 
 	// The console library examples at 'console_libraries/prom.lib' still depend on old asset files being served under `classic`.
-	router.Get("/classic/static/*filepath", func(w http.ResponseWriter, r *http.Request) {
-		r.URL.Path = path.Join("/static", route.Param(r.Context(), "filepath"))
-		fs := server.StaticFileServer(ui.Assets)
-		fs.ServeHTTP(w, r)
-	})
+	router.Get("/classic/static/*filepath", route.FileServe(filepath.Join(o.LocalAssets, "static")))
 
 	router.Get("/version", h.version)
 	router.Get("/metrics", promhttp.Handler().ServeHTTP)
@@ -443,10 +439,12 @@ func New(logger *slog.Logger, o *Options) *Handler {
 
 	router.Get("/consoles/*filepath", readyf(h.consoles))
 
-	serveReactApp := func(w http.ResponseWriter, _ *http.Request) {
-		indexPath := reactAssetsRoot + "/index.html"
-		f, err := ui.Assets.Open(indexPath)
-		if err != nil {
+	serveReactApp := func(w http.ResponseWriter, r *http.Request) {
+		f, err := os.Open(filepath.Join(o.LocalAssets, "static/react/index.html"))
+		if errors.Is(err, fs.ErrNotExist) {
+			http.ServeFile(w, r, "/usr/share/prometheus/web/no-ui/index.html")
+			return
+		} else if err != nil {
 			w.WriteHeader(http.StatusInternalServerError)
 			fmt.Fprintf(w, "Error opening React index.html: %v", err)
 			return
@@ -490,25 +488,18 @@ func New(logger *slog.Logger, o *Options) *Handler {
 
 	// The favicon and manifest are bundled as part of the React app, but we want to serve
 	// them on the root.
-	for _, p := range []string{"/favicon.svg", "/favicon.ico", "/manifest.json"} {
-		assetPath := reactAssetsRoot + p
+	for _, p := range []string{"/favicon.ico", "/manifest.json"} {
 		router.Get(p, func(w http.ResponseWriter, r *http.Request) {
-			r.URL.Path = assetPath
-			fs := server.StaticFileServer(ui.Assets)
-			fs.ServeHTTP(w, r)
+			http.ServeFile(w, r, filepath.Join(o.LocalAssets, "static/react", p))
 		})
 	}
 
-	reactStaticAssetsDir := "/assets"
-	if h.options.UseOldUI {
-		reactStaticAssetsDir = "/static"
-	}
+	// reactStaticAssetsDir := "/assets"
+	// if h.options.UseOldUI {
+	// 	reactStaticAssetsDir = "/static"
+	// }
 	// Static files required by the React app.
-	router.Get(reactStaticAssetsDir+"/*filepath", func(w http.ResponseWriter, r *http.Request) {
-		r.URL.Path = path.Join(reactAssetsRoot+reactStaticAssetsDir, route.Param(r.Context(), "filepath"))
-		fs := server.StaticFileServer(ui.Assets)
-		fs.ServeHTTP(w, r)
-	})
+	router.Get("/static/*filepath", route.FileServe(filepath.Join(o.LocalAssets, "static/react/static")))
 
 	if o.UserAssetsPath != "" {
 		router.Get("/user/*filepath", route.FileServe(o.UserAssetsPath))
